### ISSUE TITLE
ImplicitIO troubles

### ISSUE THREAD

**Issue by radeusgd** [Luna-Tensorflow/Luna-Tensorflow#91]
There are issues with highlevel functions faulting with runtime_error Cannot find storage for pointer.
This is related to Luna's lazy evaluation issues but likely can be fixed by writing our low-level wrappers correctly.

We need to investigate what can be done about this as it's very bad for end-users to get such errors.

**Comment by radeusgd** [comment]
First suspect is https://github.com/Luna-Tensorflow/Luna-Tensorflow/blob/master/API/Tensorflow/src/CWrappers/Operations.luna#L136
because the ManagedPtr may be freed before our FFI call finalizes, it should be replaced by normal Pointer (and free *after* FFI call).

We should also temporarily remove the "implicit IO safety nets" in our code to test what really crashes and fix it, so that we don't have to enforce this wierd style of coding on end users.

**Comment by Swordlash** [comment]
Although possibly prone to cause errors, it appears to me that it is not the (main) source of our problems. The ManagedPointers have been nevertheless replaced in majority of FFI calls with normal Pointers and manual free (21252f708be804b0ee7965bee14296ff4d9dcd5d). It is also possible to remove them in op generator, which uses ManagedPointers to keep attribute lists.

**Comment by Swordlash** [comment]
It seems that force-evaluating all pointer-owning objects at the beginning of each function that makes use of them is necessary to keep them allocated only once and not deallocated too early, _exempli gratia:_
```
def oper ptr1 ptr2:
    nptr1 = ptr1
    nptr2 = ptr2
   ... do something on nptrs
```

**Comment by radeusgd** [comment]
Thanks to @wojtrz13 5f43a46 and with the help of the Luna team we've found the core issue is related to the ManagedPointer interface which will be changed in new Luna version - a `with :: ManagedPointer a -> (Pointer a -> b) -> b` will be added to Luna's stdlib and using it will fix the issue, however in our internal functions we'll still want to force evaluate pointer objects as per @Swordlash comment to avoid executing the constructors multiple times needlessly.

**Comment by wojtrz13** [comment]
An example of correct handling of current ManagedPointers: on Luna side:
```
import Std.Foreign
import Std.Foreign.C.Value

def wrapNumber n:
    wrapFun = lookupSymbol "TFL" "wrap_number"
    releaseMethod = lookupSymbol "TFL" "free_number"
    ptr = wrapFun.call (Pointer CInt64) [CInt64.fromInt n . toCArg]
    managedPtr = ManagedPointer CInt64 . fromPointer releaseMethod ptr
    managedPtr

def readNumber ptr:
    ptr' = ptr
    readFun = lookupSymbol "TFL" "read_number"
    cVal = readFun.call CInt64 [ptr'.toCArg]
    val = cVal.toInt
    val

def main:
    1.upto 1000 . each x: print $ readNumber (wrapNumber 3)
```
and on C++ side:
```
TFL_API int64_t* wrap_number(int64_t number) {
    FFILOG(number);
    int64_t* ptr = static_cast<int64_t*>(malloc(sizeof(int64_t)));
    *ptr = number;
    return ptr;
}

TFL_API int64_t read_number(int64_t *ptr) {
    FFILOG(ptr);
    return *ptr;
}

TFL_API void free_number(int64_t *ptr) {
    FFILOG(ptr);
    free(ptr);
}
```

**Comment by Swordlash** [comment]
Pointers evaluation enforcement was implemented by @wojtrz13 in commit 511b69e.
